overlord/snapstate/handlers.go

```diff
@@ -4006,43 +4007,45 @@ func refreshedSnaps(reTask *state.Task) (snapNames []string, failed bool) {
 		if task.Kind() == "check-rerefresh" {
 			// we've reached a previous check-rerefresh (but not ourselves).
 			// Only snaps in tasks after this point are of interest.
-			laneSnaps = map[int]string{}
-		}
-		lanes := task.Lanes()
-		// Tasks can have multiple lanes, but can also share a single lane, right now we rely on
-		// the first lane, which may not be sufficient to detect whether or not a specific
-		// snap has been refreshed.
-		// XXX: If snaps have been refreshed with Flags.Transaction == "all-snap", then this may
-		// be an issue and not accurately detect which snaps have been refreshed.
-		lane := lanes[0]
-		if lane == 0 {
-			// not really a lane
-			continue
+			laneSnaps = make(map[int]map[string]bool)
 		}
-		if task.Status() != state.DoneStatus {
-			// ignore non-successful lane (1)
-			laneSnaps[lane] = ""
+
+		// Ignore tasks on '0' lane, they are not refreshes anyway.
+		taskLanes := task.Lanes()
+		if len(taskLanes) == 1 && taskLanes[0] == 0 {
 			continue
 		}
-		if _, ok := laneSnaps[lane]; ok {
-			// ignore lanes we've already seen (including ones explicitly ignored in (1))
-			continue
+
+		// Track lanes that failed.
+		if task.Status() != state.DoneStatus {
+			for _, l := range taskLanes {
+				failedLanes[l] = true
+			}
 		}
+
 		var snapsup SnapSetup
 		if err := task.Get("snap-setup", &snapsup); err != nil {
```

Comment:
as we are less in control about what we look at, maybe we should skip this for anything that is not a download-snap, notice that most tasks don't have directly snap-setup on them anyway but a pointer to a task that has it, so this code was  doing nothing on most tasks anyway. The tests might need changing though

Reply:
yea that's great input, this is now added